File: src/main/java/com/googongill/aditory/external/chatgpt/ChatGptService.java
package com.googongill.aditory.external.chatgpt;

import com.googongill.aditory.external.chatgpt.dto.AutoCategorizeResult;
import com.googongill.aditory.external.chatgpt.dto.ChatGptRequest;
import com.googongill.aditory.external.chatgpt.dto.ChatGptResponse;
import com.googongill.aditory.external.chatgpt.dto.Message;
import jakarta.transaction.Transactional;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

import java.util.ArrayList;
import java.util.List;

import static org.springframework.http.HttpHeaders.AUTHORIZATION;

@Slf4j
@Service
@Transactional
@RequiredArgsConstructor
public class ChatGptService {

    private final RestTemplate restTemplate = new RestTemplate();
    private final ChatGptConfig chatGptConfig;

    public HttpEntity<ChatGptRequest> buildHttpEntity(ChatGptRequest chatGptRequest) {
        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);
        headers.set("charset", "UTF-8");
        headers.add(AUTHORIZATION, "Bearer " + chatGptConfig.getApiKey());
        return new HttpEntity<>(chatGptRequest, headers);
    }

    public ChatGptResponse getResponse(HttpEntity<ChatGptRequest> chatGptRequestHttpEntity) {
        ResponseEntity<ChatGptResponse> responseEntity = restTemplate.postForEntity(
                chatGptConfig.getUrl(),
                chatGptRequestHttpEntity,
                ChatGptResponse.class);
        return responseEntity.getBody();
    }

    private static ArrayList<Message> createMessages(String messageContent) {
        ArrayList<Message> messages = new ArrayList<>();
        messages.add(Message.builder()
                .role("user")
                .content(messageContent)
                .build()
        );
        return messages;
    }

    private static String createMessageContent(String url, List<String> userCategoryNameList) {
        String categories = String.join(", ", userCategoryNameList);
        String messageContent = url + " summarizing it in one line with a 'summary', a fitting 'title', " +
                                        "and selecting the most appropriate category among " +
                                categories + ", returning it as a JSON object named 'category'.";
        log.info("message= {}", messageContent);
        return messageContent;
    }

    public AutoCategorizeResult autoCategorizeLink(String url, List<String> userCategoryNameList) {
        // message-content 생성
        String messageContent = createMessageContent(url, userCategoryNameList);
        // messages 생성
        ArrayList<Message> messages = createMessages(messageContent);
        // http entity 생성
        HttpEntity<ChatGptRequest> chatGptRequestHttpEntity = buildHttpEntity(ChatGptRequest.builder()
                .model(chatGptConfig.getModel())
                .messages(messages)
                .maxTokens(chatGptConfig.getMaxToken())
                .temperature(chatGptConfig.getTemperature())
                .topP(chatGptConfig.getTopP())
                .build());
        // response 수신
        ChatGptResponse chatGptResponse = getResponse(chatGptRequestHttpEntity);

        // 자동 분류 결과 반환
        return AutoCategorizeResult.of(chatGptResponse);
    }
}

File: src/main/java/com/googongill/aditory/external/chatgpt/ChatGptService.kt
package com.googongill.aditory.external.chatgpt

import com.googongill.aditory.external.chatgpt.dto.AutoCategorizeResult
import com.googongill.aditory.external.chatgpt.dto.ChatGptRequest
import com.googongill.aditory.external.chatgpt.dto.ChatGptResponse
import com.googongill.aditory.external.chatgpt.dto.Message
import jakarta.transaction.Transactional
import org.springframework.http.HttpEntity
import org.springframework.http.HttpHeaders
import org.springframework.http.MediaType
import org.springframework.stereotype.Service
import org.springframework.web.client.RestTemplate

@Service
//@Transactional
class ChatGptService(
    private val chatGptConfig: ChatGptConfig
) {
    private val restTemplate = RestTemplate()

    private fun buildHttpEntity(chatGptRequest: ChatGptRequest): HttpEntity<ChatGptRequest> {
        val headers = HttpHeaders().apply {
            contentType = MediaType.APPLICATION_JSON
            set("charset", "UTF-8")
            add(HttpHeaders.AUTHORIZATION, "Bearer ${chatGptConfig.apiKey}")
        }

        return HttpEntity(chatGptRequest, headers)
    }

    private fun getResponse(chatGptRequestHttpEntity: HttpEntity<ChatGptRequest>?): ChatGptResponse {
        val responseEntity = restTemplate.postForEntity(
            chatGptConfig.url,
            chatGptRequestHttpEntity,
            ChatGptResponse::class.java
        )

        return responseEntity.body
    }

    fun autoCategorizeLink(url: String, userCategoryNameList: List<String>): AutoCategorizeResult {
        // message-content 생성
        val messageContent = createMessageContent(url, userCategoryNameList)
        // messages 생성
        val messages = createMessages(messageContent)
        // http entity 생성
        val chatGptRequestHttpEntity = buildHttpEntity(
            ChatGptRequest(
                model = chatGptConfig.model,
                messages = messages,
                maxTokens = chatGptConfig.maxToken,
                temperature = chatGptConfig.temperature,
                topP = chatGptConfig.topP,
            )
        )
        // response 수신
        val chatGptResponse = getResponse(chatGptRequestHttpEntity)

        // 자동 분류 결과 반환
        return AutoCategorizeResult.of(chatGptResponse)
    }

    companion object {
        private fun createMessages(messageContent: String): List<Message> {
            return listOf(
                Message(
                    role = "user",
                    content = messageContent
                )
            )
        }

        private fun createMessageContent(url: String, userCategoryNameList: List<String>): String {
            val categories = userCategoryNameList.joinToString(", ")
            return "$url summarizing it in one line with a 'summary', a fitting 'title', " +
                    "and selecting the most appropriate category among $categories, returning it as a JSON object named 'category'."
        }
    }
}
